consensus/core/src/block_verifier.rs

```diff
@@ -1,31 +1,289 @@
+/// be accepted into the DAG.
+pub(crate) struct SignedBlockVerifier {
+    context: Arc<Context>,
+    transaction_verifier: Arc<dyn TransactionVerifier>,
+}
+
+impl SignedBlockVerifier {
+    pub(crate) fn new(
+        context: Arc<Context>,
+        transaction_verifier: Arc<dyn TransactionVerifier>,
+    ) -> Self {
+        Self {
+            context,
+            transaction_verifier,
+        }
+    }
 }
 
-#[derive(Clone)]
-pub struct TestBlockVerifier;
+// All block verification logic are implemented below.
+impl BlockVerifier for SignedBlockVerifier {
+    fn verify(&self, block: &SignedBlock) -> ConsensusResult<()> {
+        let committee = &self.context.committee;
+        // The block must belong to the current epoch and have valid authority index,
+        // before having its signature verified.
+        if block.epoch() != committee.epoch() {
+            return Err(ConsensusError::WrongEpoch {
+                expected: committee.epoch(),
+                actual: block.epoch(),
+            });
+        }
+        if !committee.is_valid_index(block.author()) {
+            return Err(ConsensusError::InvalidAuthorityIndex {
+                index: block.author(),
+                max: committee.size() - 1,
+            });
+        }
+        // Verifiy the block's signature.
+        block.verify_signature(&self.context)?;
+        // Verify the block's ancestor refs are consistent with the block's round,
+        // and total parent stakes reach quorum.
+        if block.ancestors().len() > committee.size() {
+            return Err(ConsensusError::TooManyAncestors(
+                block.ancestors().len(),
+                committee.size(),
+            ));
+        }
+        let mut own_ancestor = false;
+        let mut seen_parents = vec![false; committee.size()];
+        let mut parent_stakes = 0;
+        for ancestor in block.ancestors() {
+            if !committee.is_valid_index(ancestor.author) {
+                return Err(ConsensusError::InvalidAuthorityIndex {
+                    index: ancestor.author,
+                    max: committee.size() - 1,
+                });
+            }
+            if ancestor.author == block.author() {
+                own_ancestor = true;
+            }
+            if ancestor.round >= block.round() {
+                return Err(ConsensusError::InvalidAncestorRound {
+                    ancestor: ancestor.round,
+                    block: block.round(),
+                });
+            }
+            if ancestor.round == block.round() - 1 && !seen_parents[ancestor.author] {
+                seen_parents[ancestor.author] = true;
+                parent_stakes += committee.stake(ancestor.author);
+            }
+            // TODO: reject blocks with multiple ancestors from the same authority.
+        }
+        if !own_ancestor {
+            return Err(ConsensusError::MissingOwnAncestor);
+        }
```

Comment:
Yes, own ancestor is latest is not enforceable during verification time, since the actual latest block may not be in the causal history or received by the verifier, so verfication results on the same block can be different. If own ancestor is not latest, the best a validator can do is to discover the behavior later and report it. Downstream logic can't assum own ancestor is latest as well.

Reply:
Agreed, ok we are aligned then on this.